File: src/python/main/helper/rich_json_file_helper.py
import json
import os

from .rich_json_helper import concat_strings, merge_into_target
from ..core.rich_json import set_command_enabled, parse_rich_json
from ..core.rich_json_constants import _RICH_JSON_LATE_APPLIES
from ..other.rich_json_configuration import _RICH_JSON_CONFIG

FILE_CACHE = {}


def read_directory(path, _execute_late_applies=False):
    """
    Reads a directory like a JSON file and resolves RichJson.

    :param path: The path to the directory.
    :param _execute_late_applies: Whether to execute late applies (default: False).
    :return: A dictionary containing the resolved JSON data.
    """
    rv = {}

    if not _execute_late_applies:
        for cmd in _RICH_JSON_LATE_APPLIES:
            set_command_enabled(cmd, False)

    with os.scandir(path) as entries:
        for entry in entries:
            if entry.is_file():
                name_without_extension = os.path.splitext(entry.name)[0]
                rv[name_without_extension] = read_file(concat_strings(path, "/", entry.name), True)
            elif entry.is_dir():
                rv[entry.name] = read_directory(concat_strings(path, "/", entry.name), True)

    if not _execute_late_applies:
        for cmd in _RICH_JSON_LATE_APPLIES:
            set_command_enabled(cmd, True)

    return rv


def read_file(path, _execute_late_applies=False):
    """
    Reads a JSON file and resolves RichJson if contained.

    :param path: The path to the file.
    :param _execute_late_applies: Whether to execute late applies (default: False).
    :return: The parsed JSON data (dict or list).
    """
    if _RICH_JSON_CONFIG.get('file_cache_enabled') and path in FILE_CACHE:
        return FILE_CACHE[path]

    if _RICH_JSON_CONFIG.get('file_cache_enabled'):
        FILE_CACHE[path] = {}

    if not _execute_late_applies:
        for cmd in _RICH_JSON_LATE_APPLIES:
            set_command_enabled(cmd, False)

    with open(path, 'r', encoding='utf-8') as f:
        rv = json.load(f)

    rv = parse_rich_json(rv)

    if not _execute_late_applies:
        for cmd in _RICH_JSON_LATE_APPLIES:
            set_command_enabled(cmd, True)

    if _RICH_JSON_CONFIG.get('file_cache_enabled'):
        FILE_CACHE[path] = merge_into_target(FILE_CACHE[path], rv)

    return rv

File: src/js/main/helper/RichJsonFileHelper.js
import fs from 'fs';
import {mergeIntoTarget} from "./RichJsonHelper.js";
import {parse} from "../core/RichJson_parse.js";
import * as PATH from 'node:path';
import {__RICH_JSON_CONFIG} from "../other/RichJsonConfiguration.js";
import {__RICH_JSON_LATE_APPLIES, setCommandEnabled} from "../core/RichJsonCommandHolder.js";

const FILE_CACHE = {};

/**
 * Reads a directory like a JSON file and resolves RichJSON.
 * @param path
 * @param _executeLateApplies = false
 * @returns {{}}
 */
export function readDirectory(path, _executeLateApplies = false) {
    let entries = fs.readdirSync(path, {withFileTypes: true});
    let rv = {};

    if (!_executeLateApplies) {
        __RICH_JSON_LATE_APPLIES.forEach(cmd => setCommandEnabled(cmd, false));
    }

    entries.forEach(entry => {
        if (entry.isFile()) {
            let nameWithoutExtension = PATH.parse(entry.name).name;
            rv[nameWithoutExtension] = readFile(path + "/" + entry.name, true);
        } else if (entry.isDirectory()) {
            rv[entry.name] = readDirectory(path + "/" + entry.name, true);
        }
    });

    if (!_executeLateApplies) {
        __RICH_JSON_LATE_APPLIES.forEach(cmd => setCommandEnabled(cmd, true));
    }

    return rv;
}

/**
 * Reads a JSON file and resolves RichJSON if contained.
 * @param path
 * @param _executeLateApplies = false
 * @returns {*|string}
 */
export function readFile(path, _executeLateApplies = false) {
    if (__RICH_JSON_CONFIG.fileCacheEnabled && Object.hasOwn(FILE_CACHE, path)) {
        return FILE_CACHE[path];
    }

    if (__RICH_JSON_CONFIG.fileCacheEnabled) {
        FILE_CACHE[path] = {};
    }

    if (!_executeLateApplies) {
        __RICH_JSON_LATE_APPLIES.forEach(cmd => setCommandEnabled(cmd, false));
    }

    let rv = JSON.parse(fs.readFileSync(path, 'utf-8'));
    rv = parse(rv);

    if (!_executeLateApplies) {
        __RICH_JSON_LATE_APPLIES.forEach(cmd => setCommandEnabled(cmd, true));
    }

    if (__RICH_JSON_CONFIG.fileCacheEnabled) {
        FILE_CACHE[path] = mergeIntoTarget(FILE_CACHE[path], rv);
    }

    return rv;
}